### PR TITLE
Fix calls to np.ma.masked_where

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

Due to numpy/numpy#18967, I've audited all our calls to `np.ma.masked_*(copy=False)`. When the input comes directly from the user, or is already stored somewhere already, then we don't want these calls to modify the existing values.

This affects `np.ma.masked_where` and all _simple_ wrappers around it like `masked_less_equal`. It does _not_ affect `masked_invalid`, which continues to not modify the input. I have asked upstream https://github.com/numpy/numpy/issues/19332 whether that is intended. If they intend to change `masked_invalid` also, then we will need to make a few more changes like this, but we are mostly okay.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [n/a] New features are documented, with examples if plot related.
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/colors.py [hunk 1]
```diff
@@ -1545,11 +1545,11 @@ class LogNorm(Normalize):
 
     def autoscale(self, A):
         # docstring inherited.
-        super().autoscale(np.ma.masked_less_equal(A, 0, copy=False))
+        super().autoscale(np.ma.array(A, mask=(A <= 0)))
```
Comment (greglucas) [inline lib/matplotlib/colors.py:1548]: Does this save the copy step? It seems like this is creating a new array anyways, so I'm wondering if it is gaining anything relative to just removing the `copy=False`?

File: lib/matplotlib/tests/test_image.py [hunk 2]
```diff
@@ -1233,6 +1233,24 @@ def test_imshow_quantitynd():
     fig.canvas.draw()
 
 
+@check_figures_equal(extensions=['png'])
+def test_norm_change(fig_test, fig_ref):
+    # LogNorm should not mask anything invalid permanently.
```
Comment (greglucas) [inline lib/matplotlib/tests/test_image.py:1238]: Do you want to add in some invalids beforehand that should be kept throughout as well and not unmasked at any point?

File: lib/matplotlib/colors.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/colors.py:1548]: The default for `np.ma.array` is `copy=False, keep_mask=True`. The former will not copy the data, and the latter will use the existing mask (if any) with the new mask, which will make a copy. Removing `copy=False` from `masked_less_equal` makes a copy of both.

Comment (greglucas) [reply lib/matplotlib/colors.py:1548]: Thanks for clarifying! I was curious how to test if this actually holds and learned about `np.shares_memory` today. I did verify this assertion with that.

Comment (QuLogic) [reply lib/matplotlib/colors.py:1548]: You can also check, if you have `b = np.ma.array(a)`, that `b.base is a` (if `a` is originally a plain array), or `b.base is a.base` (if `a` is originally a masked array).

File: lib/matplotlib/tests/test_image.py (hunk 2 above)
Comment (QuLogic) [reply lib/matplotlib/tests/test_image.py:1238]: Sure, added.

**Review by greglucas — APPROVED** [review]

**Review by dstansby — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@QuLogic a small note that I _think_ this change can cause a bug when the array `A` is not an array but a list. In `np.ma.masked_less_equal(A, 0, copy=False)`, the `A` as input to the function is converted an array, while in `np.ma.array(A, mask=(A <= 0))` it already needs to be an array up front (otherwise `A <= 0` will fail). 

We ran into this with GeoPandas (https://github.com/geopandas/geopandas/pull/2066, didn't test it, but I assume this PR is the cause), where we were (incorrectly) setting the array of a color map with a list `cmap.set_array(np.array([]))`. 
This was of course wrong to do, as the docs clearly say it should be an ndarray (although in general people might expect that functions needing an array also take general "array-likes", as do most functions in numpy).

Just noting it here, I'll let it to your judgement whether this is worth addressing or not (it's certainly not needed just for GeoPandas).

**Comment by greglucas** [comment]
Just a note that I am pretty sure this is fixed in master due to a new call to a call to `cbook.safe_masked_invalid()` in `set_array()` now.
https://github.com/matplotlib/matplotlib/blob/ca275dca26d746fb1ce59a16e8c0f7db42d6813a/lib/matplotlib/cm.py#L459

v3.4.3
```python
>>> sm.set_array([])
>>> sm.get_array()
[]
```

master
```python
>>> sm.set_array([])
>>> sm.get_array()
masked_array(data=[],
             mask=False,
       fill_value=1e+20,
            dtype=float64)
```

I don't know if another 3.4.x release is planned sometime, but it should be a pretty simple fix to add that call in to cast to array immediately in the set_array call.

**Comment by QuLogic** [comment]
That came in from #18870, which I don't think we can backport fully. But we can make a small correction to this, as it's not required that `autoscale` must be called internally only.

**Pull request merged**